### PR TITLE
Add support of delegated EventListener. Fix of: #365

### PR DESCRIPTION
**Pull request by neworld**
### What does this PR do?

Add a way to provide delegate EventListener in order to solve problems mentioned in #365

### Motivation

I want to keep hidden DataDog as detail in Gradle modules and my custom EventListener is not able to use hidden DataDogEventListener. So making this PR is a more preferable solution than opening details in my project. 

### Additional Notes

Your test naming conventions are new to me. I am not sure how to specify the `when` part of all methods. 

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogEventListenerFactoryTest.kt [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @Test
+    fun `𝕄 delegate are called 𝕎 all methods`() {
+        // Given
+        val testedListener = DatadogEventListener.Factory(testedFactory).create(mockCall)
+        val address = InetSocketAddress(0)
+
+        // When
+        testedListener.callStart(mockCall)
```
Comment (mariusc83) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogEventListenerFactoryTest.kt:84]: those assertions/test methods should be moved in the `DatadogEventListenerTest` class as follows:

```
fun `M delegate W responseHeadersStart()`{
     // GIVEN
     testedListener = DatadogEventListener(fakeKey, mockDelegate)
    
    // WHEN
    testedListener.responseHeadersStart(mockCall)
   
   // THEN
   verify(mockDelegate).responseHeaderStart(mockCall)
   verifyNoMoreInteractions(mockDelegate)
}
```
There should be a similar test method for each of the `EventListener` implemented method.
The main reason behind this approach is to have each test method testing one single key feature in the SDK, this way it is easier for us to see/identify the issue whenever something goes wrong in our CI server.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogEventListener.kt [hunk 2]
```diff
@@ -211,12 +275,23 @@ internal constructor(val key: String) : EventListener() {
      *       .eventListenerFactory(new DatadogEventListener.Factory())
      *       .build();
      * ```
+     *
+     * @param delegate Optionally you could provide another EventListener.Factory as delegate witch
+     * will get all events
      */
-    class Factory : EventListener.Factory {
+    class Factory(
+        val delegate: EventListener.Factory = NoOpEventListenerFactory()
```
Comment (mariusc83) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogEventListener.kt:283]: I would name this `delegateFactory`

**Review by mariusc83 — CHANGES_REQUESTED** [review]
Thank you very much for this PR, we really appreciate you took the time to write it. We do need some changes on the Unit tests side mostly to align with our internal best practices but overall looks really good.

File: dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogEventListenerFactoryTest.kt (hunk 1 above)
Comment (neworld) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogEventListenerFactoryTest.kt:84]: My first implementation was in `DatadogEventListenerTest`, but then I changed in order to make sure delegates are working via the factory. 

> The main reason behind this approach is to have each test method testing one single key feature in the SDK, this way it is easier for us to see/identify the issue whenever something goes wrong in our CI server.

In this case, would be more practical to write a dynamic test that collects all methods and make sure no one is missed.

Comment (mariusc83) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogEventListenerFactoryTest.kt:84]: I am not sure how this dynamic test would look like but if it is going to have a dedicated test method for delegated method inside the EventListener that's the way. I know it is a lot of boiler code but it is better if you want to have relevant status messages from your CI server.

Comment (mariusc83) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogEventListenerFactoryTest.kt:84]: Regarding the factory, its only responsibility is to build that instance (DatadogEventListener in this case). You don't have its business logic, the most you have to do is to test that it really contains that delegate. The business logic is going to be tested by the `DatadogEventListenerTest` class.

**Comment by neworld** [comment]
Suddenly I come to the realization this solution does not cover all cases and actually it could be even evil. [EventListener class gains more methods](https://github.com/square/okhttp/blob/master/CHANGELOG.md#user-content-version-470) other time. This PR covers only methods included in 3.12.6. Also, this approach leads to the wrong assumptions.

**Comment by mariusc83** [comment]
> Suddenly I come to the realization this solution does not cover all cases and actually it could be even evil. [EventListener class gains more methods](https://github.com/square/okhttp/blob/master/CHANGELOG.md#user-content-version-470) other time. This PR covers only methods included in 3.12.6. Also, this approach leads to the wrong assumptions.

@neworld in this case you might go the other way around and wrap our DatadogEventListener in your own implementation and delegate all the methods to ours. This way you will be able to maintain yours and indeed it will avoid the false assumptions that all the methods will be delegated to the delegate.

**Comment by neworld** [comment]
> This way you will be able to maintain yours and indeed it will avoid the false assumptions that all the methods will be delegated to the delegate.

Yes, indeed. This way will be safer as I got more control.

**Pull request closed without merging**